src/QuadForm/Torsion.jl

```diff
@@ -733,17 +759,383 @@ function preimage(f::TorQuadModMor, a::TorQuadModElem)
+"""
 is_bijective(f::TorQuadModMor) = is_bijective(f.map_ab)
 
+@doc Markdown.doc"""
+    is_surjective(f::TorQuadModMor) -> Bool
+
+Return whether `f` is surjective.
+"""
+is_surjective(f::TorQuadModMor) = is_surjective(f.map_ab)
+
+@doc Markdown.doc"""
+    is_injective(f::TorQuadModMor) -> Bool
+
+Return whether `f` is injective.
+"""
+is_injective(f::TorQuadModMor) = is_injective(f.map_ab)
+
+# Rely on the algorithm implemeted for GrpAbFinGenMap
+@doc Markdown.doc"""
+    has_complement(i::TorQuadModMor) -> Bool, TorQuadModMor
+
+Given a map representing the injection of a submodule $W$ of a torsion
+quadratic module $T$, return whether $W$ has an orthogonal complement
+$U$ in $T$. If yes, it returns an injection $U \to T$.
+"""
+function has_complement(i::TorQuadModMor)
+  @req is_injective(i) "i must be injective"
+  T = codomain(i)
+  bool, jab = Hecke.has_complement(i.map_ab)
+  if !bool
+    return (false, sub(T, TorQuadModElem[])[2])
+  end
+  Qab = domain(jab)
+  Q, j = sub(T, [T(jab(a)) for a in gens(Qab)])
+  @assert is_injective(j)
+  return (true, j)
+end
+
 ################################################################################
 #
-#  Submodules
+#  (Anti)-Isometry
 #
 ################################################################################
 
+# we test isometry in the semi-regular case: we compare the gram matrices of the
+# quadratic forms associated to the respective normal forms.
+function _isometry_semiregular(T::TorQuadMod, U::TorQuadMod)
+  # the zero map for default output
+  hz = hom(T, U, zero_matrix(ZZ, ngens(T), ngens(U)))
+  NT, TtoNT = normal_form(T)
+  NU, UtoNU = normal_form(U)
+  gqNT = gram_matrix_quadratic(NT)
+  gqNU = gram_matrix_quadratic(NU)
+  if gqNT != gqNU
+    return (false, hz)
+  end
+  NTtoNU = hom(NT, NU, identity_matrix(ZZ, ngens(NT)))
+  TtoU = compose(TtoNT, compose(NTtoNU, inv(UtoNU)))
+  @assert is_bijective(TtoU)
+  @assert all(a -> a*a == TtoU(a)*TtoU(a), gens(T))
+  return (true, TtoU)
+end
+
+# we test in the degenerate case. For now, we only cover the case where both T and U
+# split into a direct sum of their respective quadratic radical. If not, we return
+# "Not yet implemented". If yes, we compare the normal forms of the respective complements
+# which are non-degenerate, and if the radicals have the same elementary divisors, we
+# complete the isometry by adding the identity matrix from one radical to the other one.
+function _isometry_degenerate(T::TorQuadMod, U::TorQuadMod)
+  # the zero map for default output
+  hz = hom(T, U, zero_matrix(ZZ, ngens(T), ngens(U)))
+  rqT, rqTtoT = radical_quadratic(T)
+  rqU, rqUtoU = radical_quadratic(U)
+  if elementary_divisors(rqT) != elementary_divisors(rqU)
+    return (false, hz)
+  end
+  # at this point we can map safely one radical to the other one
+  boolT, jT = has_complement(rqTtoT)
+  boolU, jU = has_complement(rqUtoU)
+  @req boolT && boolU "Not yet implemented" # Non split case
+  NT = domain(jT)
```

Comment:
shouldn't this rather be the image?

Reply:
No, the way it is constructed, `jT` defines the embedding of its domain into its codomain